Sources/VitalHealthKit/HealthKit/VitalHealthKitClient.swift

```diff
@@ -196,27 +196,51 @@ extension VitalHealthKitClient {
     self.backgroundDeliveryTask?.cancel()
     
     let allowedSampleTypes = Set(resources.flatMap(toHealthKitTypes(resource:)))
-    let common = Set(observedSampleTypes()).intersection(allowedSampleTypes)
-    let sampleTypes = common.compactMap { $0 as? HKSampleType }
-    
-    if sampleTypes.isEmpty {
+
+    let set: [Set<HKObjectType>] = observedSampleTypes().map(Set.init)
+    let common: [[HKSampleType]] = set.map { $0.intersection(allowedSampleTypes) }.map { $0.compactMap { $0 as? HKSampleType } }
+    let cleaned: Set<[HKSampleType]> = Set(common.filter { $0.isEmpty == false })
+
+    let uniqueFlatenned: Set<HKSampleType> = Set(cleaned.reduce([]) { acc, next in
+      var copy = acc
+      copy.append(contentsOf: next)
+      return copy
+    })
+
+    if uniqueFlatenned.isEmpty {
       logger?.info("Not observing any type")
     }
-    
+
     /// Enable background deliveries
-    enableBackgroundDelivery(for: sampleTypes)
-    
-    let stream = backgroundObservers(for: sampleTypes)
-    
+    enableBackgroundDelivery(for: uniqueFlatenned)
+
+    let stream: AsyncStream<BackgroundDeliveryPayload>
+
+    if #available(iOS 15.0, *) {
+      stream = bundledBackgroundObservers(for: cleaned)
+    } else {
+      stream = backgroundObservers(for: uniqueFlatenned)
+    }
+
     self.backgroundDeliveryTask = Task(priority: .high) {
       for await payload in stream {
-        /// Sync types one-by-one
-        await sync(payload: .type(payload.sampleType), completion: payload.completion)
+
+        guard payload.sampleTypes.isEmpty == false else {
```

Comment:
[nit] `guard let firstSampleType = payload.sampleTypes.first` implies this same non-empty check & avoids force unwrap down the line.